### ISSUE TITLE
partition_entry::apply might update app_stats.row_writes after new_version->partition is moved

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#18270]
Commit 59fbbb993fb added mutation_application_stats.
In partition_entry::apply, row_writes was updated on the error handling path,
if `apply_monotonically` threw an exception.

However, after a4e96960b8109a07953df9fe34959fd51fa35233, we get to this path also
when `apply_monotonically` returned early with `stop_iteration::no`,
in which case the partition was partially applied, updating app_stats.row_writes while at it,
and since `new_version->partition()` is already moved, it's in an undetermined state.
In any case, even if it could be accessed safely after apply_monotonically preemption,
updating `app_stats.row_writes += new_version->partition().row_count();` seems wrong.

**Comment by bhalevy** [comment]
Cc @piodul

**Comment by bhalevy** [comment]
@michoecho can you please look into this?
/cc @paszkow

**Comment by bhalevy** [comment]
> > and since new_version->partition() is already moved, it's in an undetermined state
> 
> It isn't moved. `std::move` is called on it, but this doesn't translate to any move constructors/assignments. `apply_monotonically` just expects an rvalue reference as the argument, but it leaves it in a well-defined state. It only moves away _pieces_ of the partition.

You're right.
But this interface is quite confusing and it's unclear from the documentation that `p` that's passed by rvalue reference is still valid when stop_iteration::no is returned.

How about changing the signature to accept `mutation_partition_v2&` rather than `&&`?
@tgrabiec WDYT?

> > In any case, even if it could be accessed safely after apply_monotonically preemption,
> > updating `app_stats.row_writes += new_version->partition().row_count();` seems wrong.
> 
> Why? It looks fine to me.

Is some rows were applied monotonically onto `_version->partition()` they wouldn't be accounted, no?
Shouldn't we count them before calling `apply_monotonically`?
Also, if `apply_monotonically` returns `stop_iteration::yes` we don't update `row_writes` at all.
Is that intentional?

Maybe there's something I'm missing here...

> 
> The only problem with it is that it could cause a reactor stall if `new_version` is big, because `row_count` iterates over the entire tree. (But it's not expected to be big in practice, and we never witnessed that).

**Comment by michoecho** [comment]
> > > and since new_version->partition() is already moved, it's in an undetermined state
> > 
> > 
> > It isn't moved. `std::move` is called on it, but this doesn't translate to any move constructors/assignments. `apply_monotonically` just expects an rvalue reference as the argument, but it leaves it in a well-defined state. It only moves away _pieces_ of the partition.
> 
> it's unclear from the documentation that `p` that's passed by rvalue reference is still valid when stop_iteration::no is returned.

The documentation says:
```
    // Applies p to this instance.
    //
    // Monotonic exception guarantees. In case of exception the sum of p and this remains the same as before the exception.
    //
    // Returns stop_iteration::no if the operation was preempted before finished, and stop_iteration::yes otherwise.
    // On preemption the sum of this and p stays the same (represents the same set of writes), and the state of this
    // object contains at least all the writes it contained before the call (monotonicity). It may contain partial writes.
    // Also, some progress is always guaranteed (liveness).
    //
    // If returns stop_iteration::yes, then the sum of this and p is NO LONGER the same as before the call,
    // the state of p is undefined and should not be used for reading.
    //
    // The operation can be driven to completion like this:
    //
    //   apply_resume res;
    //   while (apply_monotonically(..., is_preemtable::yes, &res) == stop_iteration::no) { }
```

I think it's pretty clear.

> How about changing the signature to accept mutation_partition_v2& rather than &&?
> @tgrabiec WDYT?

Why, though? You are saying that & is better than && because the argument might still be valid after the call. But this reasoning is backwards. && doesn't exist to guarantee that the argument will be invalidated, but to guarantee that lvalues won't be invalidated. (One could argue that this is a matter of opinion/convention, but I disagree, because the latter property is actually enforced by the language, while the former convention is only a convention). So if you have a function that *can* invalidate something, it's more appropriate for it to use &&.

In other words, it's better to risk having code that does the right thing but looks like it uses invalidated objects than to risk having code which uses invalidated objects but looks okay.

> Is some rows were applied monotonically onto _version->partition() they wouldn't be accounted, no?
Shouldn't we count them before calling apply_monotonically? Also, if apply_monotonically returns stop_iteration::yes we don't update row_writes at all.
Is that intentional?
> Maybe there's something I'm missing here...

They would be accounted. You are missing the fact that `apply_monotonically` updates `row_writes` too.

At the end of `apply`, `app_stats.row_writes` will be incremented by the number of rows in `mp` regardless of when `apply_monotonically` was preempted.

**Comment by bhalevy** [comment]
> > > > and since new_version->partition() is already moved, it's in an undetermined state
> > > 
> > > 
> > > It isn't moved. `std::move` is called on it, but this doesn't translate to any move constructors/assignments. `apply_monotonically` just expects an rvalue reference as the argument, but it leaves it in a well-defined state. It only moves away _pieces_ of the partition.
> > 
> > 
> > it's unclear from the documentation that `p` that's passed by rvalue reference is still valid when stop_iteration::no is returned.
> 
> The documentation says:
> 
> ```
>     // Applies p to this instance.
>     //
>     // Monotonic exception guarantees. In case of exception the sum of p and this remains the same as before the exception.
>     //
>     // Returns stop_iteration::no if the operation was preempted before finished, and stop_iteration::yes otherwise.
>     // On preemption the sum of this and p stays the same (represents the same set of writes), and the state of this
>     // object contains at least all the writes it contained before the call (monotonicity). It may contain partial writes.
>     // Also, some progress is always guaranteed (liveness).
>     //
>     // If returns stop_iteration::yes, then the sum of this and p is NO LONGER the same as before the call,
>     // the state of p is undefined and should not be used for reading.
>     //
>     // The operation can be driven to completion like this:
>     //
>     //   apply_resume res;
>     //   while (apply_monotonically(..., is_preemtable::yes, &res) == stop_iteration::no) { }
> ```
> 
> I think the doc is pretty clear about the fact that `p` remains valid until `stop_iteration::yes`. (That's what `monotonically` is all about).
> 
> > How about changing the signature to accept mutation_partition_v2& rather than &&?
> > @tgrabiec WDYT?
> 
> Why, though? You are saying that & is better than && for this function because the argument might still be valid after the call. But this reasoning is backwards. && doesn't exist to guarantee that the argument will be invalidated, but to guarantee that an lvalue argument won't be accidentally invalidated. (One could argue that this is a matter of opinion/convention, but I disagree, because the latter property is actually enforced by the language, while the former convention can at most be only a convention). So if you have a function that _can_ invalidate something, it's more appropriate for it to use &&.
> 
> In other words, it's better to design the API in a way that risks having correct code which looks wrong, than one which risks having incorrect code which looks fine.
> 
> > Is some rows were applied monotonically onto _version->partition() they wouldn't be accounted, no?
> > Shouldn't we count them before calling apply_monotonically? Also, if apply_monotonically returns stop_iteration::yes we don't update row_writes at all.
> > Is that intentional?
> > Maybe there's something I'm missing here...
> 
> They would be accounted. You are missing the fact that `apply_monotonically` increments `row_writes` for every row it moves out of `mp`.
> 
> At the end of `apply`, `app_stats.row_writes` will be incremented by the number of rows in `mp` regardless of when `apply_monotonically` was preempted.

I see. thanks.
So this issue can be closed as invalid.
